### ISSUE TITLE
Rest Parameters Support

### ISSUE THREAD

**Issue by delchev** [mozilla/rhino#652]
Hi,

Has anyone already explored how we can support the "rest parameters" in functions?

```javascript
function sum(...theArgs) {
  return theArgs.reduce((previous, current) => {
    return previous + current;
  });
}

console.log(sum(1, 2, 3));
// expected output: 6

console.log(sum(1, 2, 3, 4));
// expected output: 10
```

> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters

Any guidance will be very much appreciated!

Regards,
Nedelcho

**Comment by p-bakker** [comment]
This case is blocking us from updating [test262](https://github.com/tc39/test262/tree/f94fc660cc3c59b1f2f9f122fc4d44b4434b935c) to their latest master, as the [harness code](https://github.com/tc39/test262/tree/f94fc660cc3c59b1f2f9f122fc4d44b4434b935c/harness) used by the tests in test262 make use of rest parameters, which crashes the entire test suite.

**Comment by p-bakker** [comment]
Support for ...restParams on (regular, so not arrow) functions has been introduced in https://github.com/mozilla/rhino/pull/1451 and available in V1.7.15

**Comment by anivar** [comment]
## Testing Results

I tested rest parameters in Rhino 1.8.1-SNAPSHOT with VERSION_ES6 (200):

```javascript
function sum(...theArgs) {
    console.log("Args:", theArgs);
    return theArgs.reduce((a, b) => a + b);
}

console.log("Rest parameters test:", sum(1, 2, 3, 4));
```

**Result:** ✅ Works perfectly! 
- Outputs `Args: [1,2,3,4]` 
- Outputs `Rest parameters test: 10`

**Code Evidence:** 
- Token.java has `DOTDOTDOT` token for spread/rest syntax
- Parser.java handles rest parameters in function definitions
- Arrow functions with rest parameters also work

Rest parameters are fully functional in current Rhino. This issue can be closed.

**Comment by rbri** [comment]
@anivar as the one that did the ompl for ordinary functions, i will have a look at the arrow function part of this sory and hopefully close this. Thanks for the reminder and your investigation.

**Comment by anivar** [comment]
@rbri  You're right.  arrow function rest parameters don't work, only regular function rest parameters do. The issue should stay open.
